lib/regexp.mjs

```diff
@@ -0,0 +1,37 @@
+/*
+  Copyright 2023 Google LLC
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+const CLASS_MATCHER = '(?:[A-Z][a-z]+)+';
+
+// Captures `class` in `@google-cloud/foo!google.api.Http:class`
+export const KIND_REGEXP = /\:([a-z]+)$/;
+
+// Captures ` (Class)` in `Http (Class)`
+export const TITLE_KIND_REGEXP = / \([A-Za-z]+\)$/;
+
+// Like `v2` or `v2beta` or `v1alpha4`
+const VERSION_ID_MATCHER = 'v[0-9]+(?:[a-z]+[0-9]*)?';
+export const VERSION_ID_REGEXP = new RegExp(VERSION_ID_MATCHER);
+
+// Like `proto` or `someFunction`
+const OTHER_SYMBOL_MATCHER = '[a-z]+' + CLASS_MATCHER;
+// Matches `v2`, `someFunction`, and `PackageClient`; captures `PackageClient`
+const ANY_SYMBOL_MATCHER = `(?:${VERSION_ID_MATCHER}|${OTHER_SYMBOL_MATCHER}|(${CLASS_MATCHER}))`;
+// Matches `v2, someFunction, PackageClient`; captures `PackageClient`
+const EXPORTS_MATCHER = `(?:${ANY_SYMBOL_MATCHER}, )*${ANY_SYMBOL_MATCHER}`;
```

Comment:
Ugh, I failed to consider non-autogenerated libraries. Sadness!

In light of that fact, I think it would be best to switch to an approach that involves actually parsing the code. That would make things less brittle all around; for example, that approach would be more resilient against minor changes to `@google-cloud/gapic-generator` that prevent regexps from matching. (It would also avoid the need for two separate code paths—one for autogenerated libraries, another for manually written libraries.) The only reason I didn't do it this way initially is that my somewhat hacky, regexp-centric approach seemed adequate at the time.

My hope is that I can get the necessary info from the generated `foo.api.json` file, either by loading it with [`@microsoft/api-extractor-model`](https://github.com/microsoft/rushstack/tree/main/libraries/api-extractor-model) or by examining it directly. If that doesn't pan out, I should be able to fall back on TypeScript's [Compiler API](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API) to create an AST, then traverse it to get the required info.

Does that sound good? If not, how do you think I should proceed? Happy to chat if you like!

Reply:
I'm not particularly against having two code paths, but I agree it is more brittle than looking at the API/source code. Traversing the AST seems a bit complex -- but I like you're suggestion of using [`@microsoft/api-extractor-model`](https://github.com/microsoft/rushstack/tree/main/libraries/api-extractor-model):

We can go through each `apiMember` under the `EntryPoint`.  From there, we can treat each `apiItem` of kind `Class` or `Enum` directly under the `EntryPoint` as top-level items in the TOC. Other `Class`es `Enum`s and `Interface`s can then be put under their respective directories in the TOC.

Alternatively, we can also look at the `.d.ts` rollup file that gets generated when we run `api-extractor`, and try to grab the exports from there. I think this is a harder option, but at least we'll have an approach consistent for auto-generated and manually written libraries.

Reply:
Sounds good! I'll give `@microsoft/api-extractor-model` a shot and let you know when I have something worth looking at (or when I get stuck).

Reply:
We're now using `@microsoft-api-extractor-model`.

We could probably use it a bit more heavily in the future. For example, rather than using a regexp to extract the namespace from the UID, we could first walk the model and create a map of UIDs to nodes; then, to find the namespace for a UID, we could get the node for that UID and walk up its ancestors until we find a namespace node. That might be more robust than the regexp approach. For now, though, the regexp seems to get the job done.